apps/passport-client/components/shared/cards/DevconnectTicket.tsx

```diff
@@ -1,41 +1,142 @@
-    return verificationLink;
-  }, [pcd]);
+    if (zk) {
+      const serializedTicketPCD = await EdDSATicketPCDPackage.serialize(pcd);
+      const serializedIdentityPCD = await SemaphoreIdentityPCDPackage.serialize(
+        pcds.getPCDsByType(SemaphoreIdentityPCDPackage.name)[0]
+      );
+      const zkPCD = await ZKEdDSAEventTicketPCDPackage.prove({
+        ticket: {
+          value: serializedTicketPCD,
+          argumentType: ArgumentTypeName.PCD
+        },
+        identity: {
+          value: serializedIdentityPCD,
+          argumentType: ArgumentTypeName.PCD
+        },
+        fieldsToReveal: {
+          value: {
+            revealEventId: true,
+            revealProductId: true,
+            revealTicketId: true
+          },
+          argumentType: ArgumentTypeName.ToggleList
+        },
+        validEventIds: {
+          value: [pcd.claim.ticket.eventId],
+          argumentType: ArgumentTypeName.StringArray
+        },
+        externalNullifier: {
+          value: undefined,
+          argumentType: ArgumentTypeName.BigInt
+        },
+        watermark: {
+          value: Date.now().toString(),
+          argumentType: ArgumentTypeName.BigInt
+        }
+      });
+      const serializedZKPCD =
+        await ZKEdDSAEventTicketPCDPackage.serialize(zkPCD);
+      const verificationLink = makeTicketIdPCDVerifyLink(
+        encodeQRPayload(JSON.stringify(serializedZKPCD))
+      );
+      return verificationLink;
+    } else {
+      const ticketId = pcd.claim.ticket.ticketId;
+      const verificationLink = makeTicketIdVerifyLink(ticketId);
+      return verificationLink;
+    }
+  }, [pcd, pcds, zk]);
 
-  return (
-    <QRDisplayWithRegenerateAndStorage
-      generateQRPayload={generate}
-      maxAgeMs={1000 * 60}
-      uniqueId={pcd.id}
-      fgColor={getQRCodeColorOverride(pcd)}
-    />
-  );
+  if (zk) {
+    return (
+      <QRDisplayWithRegenerateAndStorage
+        // Key is necessary so that React notices that this isn't the non-ZK
+        // QR code component.
+        key={`zk-${pcd.id}`}
+        generateQRPayload={generate}
+        loadingLogo={
+          <LoadingIconContainer>
+            <LoadingIcon src={icons.qrCenterLoading} />
+          </LoadingIconContainer>
+        }
+        maxAgeMs={1000 * 60}
+        // QR codes are cached by ID, so we need to distinguish the ZK version
+        // by this prefix.
+        uniqueId={`zk-${pcd.id}`}
+        fgColor={getQRCodeColorOverride(pcd)}
+      />
+    );
+  } else {
+    return (
+      <QRDisplayWithRegenerateAndStorage
+        key={pcd.id}
+        generateQRPayload={generate}
+        maxAgeMs={1000 * 60}
+        uniqueId={`${pcd.id}`}
+        fgColor={getQRCodeColorOverride(pcd)}
+      />
+    );
+  }
 }
 
 export function DevconnectCardBody({ pcd }: { pcd: EdDSATicketPCD }) {
+  const [zk, setZk] = useState<boolean>(false);
+  const onToggle = useCallback(() => {
+    setZk(!zk);
+  }, [zk]);
```

Comment:
this is really funny in the absence of any context. ZK ON! ZK OFF! I wish it were that simple...